src/wasmtime/wasmtime.cc

```diff
@@ -0,0 +1,674 @@
+  uint64_t getMemorySize() override;
+  std::optional<std::string_view> getMemory(uint64_t pointer, uint64_t size) override;
+  bool setMemory(uint64_t pointer, uint64_t size, const void *data) override;
+  bool getWord(uint64_t pointer, Word *word) override;
+  bool setWord(uint64_t pointer, Word word) override;
+
+#define _REGISTER_HOST_FUNCTION(T)                                                                 \
+  void registerCallback(std::string_view module_name, std::string_view function_name, T,           \
+                        typename ConvertFunctionTypeWordToUint32<T>::type f) override {            \
+    registerHostFunctionImpl(module_name, function_name, f);                                       \
+  };
+  FOR_ALL_WASM_VM_IMPORTS(_REGISTER_HOST_FUNCTION)
+#undef _REGISTER_HOST_FUNCTION
+
+#define _GET_MODULE_FUNCTION(T)                                                                    \
+  void getFunction(std::string_view function_name, T *f) override {                                \
+    getModuleFunctionImpl(function_name, f);                                                       \
+  };
+  FOR_ALL_WASM_VM_EXPORTS(_GET_MODULE_FUNCTION)
+#undef _GET_MODULE_FUNCTION
+private:
+  template <typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                void (*function)(void *, Args...));
+
+  template <typename R, typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                R (*function)(void *, Args...));
+
+  template <typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<void(ContextBase *, Args...)> *function);
+
+  template <typename R, typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<R(ContextBase *, Args...)> *function);
+
+  std::string source_;
+  WasmStorePtr store_;
+  WasmModulePtr module_;
+  WasmSharedModulePtr shared_module_;
+  WasmInstancePtr instance_;
+
+  WasmMemoryPtr memory_;
+  WasmTablePtr table_;
+
+  std::unordered_map<std::string, FuncDataPtr> host_functions_;
+  std::unordered_map<std::string, WasmFuncPtr> module_functions_;
+};
+
+bool Wasmtime::load(const std::string &code, bool allow_precompiled) {
+  source_ = code;
+  store_ = WasmStorePtr(wasm_store_new(engine()));
+  // Wasm file header is 8 bytes (magic number + version).
+  static const uint8_t magic_number[4] = {0x00, 0x61, 0x73, 0x6d};
+  if (source_.size() < 8 || ::memcmp(source_.data(), magic_number, 4) != 0) {
+    return false;
+  }
+
+  wasm_byte_vec_t source_vec;
+  wasm_byte_vec_new_uninitialized(&source_vec, source_.size());
+  ::memcpy(source_vec.data, source_.data(), source_.size());
+
+  wasm_module_t *module_ptr;
+  assert(wasmtime_module_new(engine(), &source_vec, &module_ptr) == nullptr);
```

Comment:
Anything inside `assert()` will be removed in the release builds, so you need to split the call and `assert()` on the result, otherwise the module will be never created.

But you should return `false` in case of failure here, instead of `assert()`. See the V8 code.